### PR TITLE
Implement a testing infrastructure

### PR DESCRIPTION
**Pull request by EmosewaMC**
Implement a proper testing infrastructure to be used with future PRs for testing and to allow for better mocks to be made.

Tests will now use the gtest to allow for test fixtures and better tests since we are now able to run tests on every file since the linking issues are now resolved.

Tests are disabled by default since general users should not need them.

CMake changes in the 4 server CMake's is to allow these libraries to be linked in tests.

Added flag sets in the CMakePresets such that tests are always run for CI.

Fixed Windows CI builds not actually being able to run DLU or tests since the libs were in the incorrect directory.

### CODE REVIEW

**Comment by EmosewaMC** [comment]
I have also added a DestroyableComponent Test as a proof of concept of what we can make tests look like.

File: dGame/dComponents/DestroyableComponent.cpp [hunk 1]
```diff
@@ -137,26 +137,15 @@ void DestroyableComponent::Serialize(RakNet::BitStream* outBitStream, bool bIsIn
...
             }
         }
-
 		m_DirtyHealth = false;
     }
 
-    if (m_DirtyThreatList || bIsInitialUpdate) {
-        outBitStream->Write1();
-        outBitStream->Write(m_HasThreats);
-		m_DirtyThreatList = false;
```
Comment (aronwk-aaron) [inline dGame/dComponents/DestroyableComponent.cpp:156]: do we need to set this back to false?

Comment (EmosewaMC) [reply dGame/dComponents/DestroyableComponent.cpp:155]: Does it _need_ to?  no.  This value is always false and never changes in the source code atm.  For brevity I will set it to false anyways.

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -26,9 +39,5 @@ configure_file(
...
 )
 
-# Add all the ADD_TEST for each test
-foreach (test ${TestsToRun})
-	get_filename_component (TName ${test} NAME_WE)
-	add_test (NAME ${TName} COMMAND CommonCxxTests ${TName})
-	set_property(TEST ${TName} PROPERTY ENVIRONMENT CTEST_OUTPUT_ON_FAILURE=1)
-endforeach ()
+include(GoogleTest)
+gtest_discover_tests(CommonCxxTests)
```
Comment (Xiphoseer) [inline tests/CMakeLists.txt:43]: Didn't know about `gtest_discover_tests`, that's neat!

**Review by Xiphoseer — CHANGES_REQUESTED** [review]
Ok, overall agree with the direction this is going. You do a ton of manually throwing around paths in cmake though (OUTPUT_DIR, foreach ... set loops), seemingly just to link everything into one giant test binary.

If possible, try to do the `install(GTest)` once at the start, then use `add_subdirectory` as needed and let every nested `CMakeLists.txt` build its own test executable linking just the stuff it should have available and then passing that to gtest discover.

Reason being, if we want one giant cmake construct to build the whole repo, we can just keep it as a single file. But if we do split it up (which I think we should, because it's easier to maintain), we should'd reach through these directories and abstractions all the time.

**Comment by Xiphoseer** [comment]
One other thing: is it necessary to disable compiling tests by default / make it optional? Can we add a comment on why do that or not?

File: CMakePresets.json [hunk 3]
```diff
@@ -1,126 +1,122 @@
...
-      "execution": {
+    ],
+    "buildPresets": [
+      {
+        "name": "default",
+        "configurePreset": "default",
+        "displayName": "Default Build",
+        "description": "Default Build",
         "jobs": 2
       },
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply CMakePresets.json:95]: The parallel job execution is not allowed due to raknet not allowing a server to be run on the same port that is already in use.  One test does not know when the other finishes and so this causes issues when running tests in parallel.

File: dAuthServer/CMakeLists.txt [hunk 4]
```diff
@@ -1,4 +1,2 @@
-set(DAUTHSERVER_SOURCES "AuthServer.cpp")
-
-add_executable(AuthServer ${DAUTHSERVER_SOURCES})
+add_executable(AuthServer "AuthServer.cpp")
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dAuthServer/CMakeLists.txt:1]: This is on theme with how I changed the other server sources the same way so that we can test them.  Would be better to keep it consistent instead of having it be different for this one because it wasn't on theme for a PR.

File: dChatServer/CMakeLists.txt [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
-target_link_libraries(ChatServer ${COMMON_LIBRARIES} dChatFilter)
+set(DCHATSERVER_SOURCES
+	"ChatPacketHandler.cpp"
+	"PlayerContainer.cpp"
+)
+
+add_executable(ChatServer "ChatServer.cpp")
+add_library(dChatServer ${DCHATSERVER_SOURCES})
+
+target_link_libraries(dChatServer ${COMMON_LIBRARIES} dChatFilter)
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply dChatServer/CMakeLists.txt:9]: While it is small, it allows us to link this library for testing in the future

File: tests/CMakeLists.txt [hunk 6]
```diff
@@ -1,19 +1,32 @@
...
+endforeach ()
+
+add_subdirectory(dGameTests)
+foreach (file ${DGAMETEST_SOURCES})
+	set(Tests ${Tests} "dGameTests/${file}")
+endforeach ()
+
+# add the executable
+add_executable(CommonCxxTests ${Tests})
+target_link_libraries(CommonCxxTests ${COMMON_LIBRARIES} GTest::gtest_main dGame dZoneManager dPhysics detour recast tinyxml2 dWorldServer dChatFilter)
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply tests/CMakeLists.txt:29]: Yeah the test name here should be updated since its no longer CommonCxxTests but currently is all tests.  Not sure if we can have multiple test exes but ill see about splitting this up a bit more

**Comment by EmosewaMC** [comment]
> One other thing: is it necessary to disable compiling tests by default / make it optional? Can we add a comment on why do that or not?

To me, if I was a generic player, I dont want to spend time compiling tests that will never need to be run.  My experience with MariaDB was that it had multiple tests that took upwards of 2 minutes to compile and since I wouldn't be running tests, it was better to just disable them.  This provides that option without dirtying the tree

File: CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xiphoseer) [reply CMakePresets.json:95]: That's a bad reason to disable that in runners completely though. If you need a free port, bind to `0` and it'll auto assign one for you.

File: CMakeVariables.txt [hunk 7]
```diff
@@ -15,6 +15,8 @@ __dynamic=1
 # __include_backtrace__=1
 # Set __include_backtrace__ to 1 to includes the backtrace library for better crashlogs.
 # __compile_backtrace__=1
-# Set __compile_backtrace__ to 1 to compile the backtrace library instead of using system libraries. 
+# Set __compile_backtrace__ to 1 to compile the backtrace library instead of using system libraries.
 __maria_db_connector_compile_jobs__=1
 # Set to the number of jobs (make -j equivalent) to compile the mariadbconn files with.
+# __enable_testing__=1
+# When set to 1 and uncommented, compiling and linking testing folders and libraries will be done.
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply CMakeVariables.txt:22]: 100% agree.  Just following the format here but can change if its wanted

File: CMakePresets.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply CMakePresets.json:95]: Interesting.  Raknet does seem to support that.  Noted and applied

[the remark this answers is not in the document]
Comment (Xiphoseer) [reply CMakePresets.json:95]: It's not a RakNet specific thing, I think all udp socket implementations support that. RN just passes it along.

File: tests/dGameTests/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+add_subdirectory(dComponentsTests)
+set(DGAMETEST_SOURCES
+	"GameDependencies.cpp"
+)
+
+foreach(file ${DCOMPONENTS_TESTS})
+	set(DGAMETEST_SOURCES ${DGAMETEST_SOURCES} "dComponentsTests/${file}")
```
[the remark this answers is not in the document]
Comment (Jettford) [reply tests/dGameTests/CMakeLists.txt:7]: Oh, I wish I knew of this when I refactored the CMake solution. Would have made the dGame CMake file much cleaner

File: CMakeLists.txt [hunk 9]
```diff
@@ -74,21 +74,30 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
 set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
 
+# Make sure to move required DLLs to the actual directory where the binaries will be!
+# For CMAKE_BUILD_TYPES this is not the binary dir.
+
+set(OUTPUT_DIR ${PROJECT_BINARY_DIR})
+if(${CMAKE_GENERATOR} MATCHES "Visual Studio 17 2022")
```
[the remark this answers is not in the document]
Comment (EmosewaMC) [reply CMakeLists.txt]: i've partially resolved this by overriding the RelWithDebInfo directory with the binary directory.  This is what I found is supported by CMake and is much cleaner than this and is actually a supported way of getting the binaries to the right folder.  Should i add the same thing for the other build types (debug, release, relminsize etc.)?

[the remark this answers is not in the document]
Comment (EmosewaMC) [reply CMakeLists.txt]: I will be making an issue to note the build type overriding but for now just overriding the output directories with
```set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELWITHDEBINFO ${CMAKE_BINARY_DIR})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELWITHDEBINFO ${CMAKE_BINARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO ${CMAKE_BINARY_DIR})
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_BINARY_DIR})
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_RELEASE ${CMAKE_BINARY_DIR})
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE ${CMAKE_BINARY_DIR})
```
ensures the build works on Windows and gets us off the ground here.

File: README.md [hunk 10]
```diff
@@ -42,6 +42,10 @@ This was done make sure that older and incomplete clients wouldn't produce false
 
 If you're using a DLU client you'll have to go into the "CMakeVariables.txt" file and change the NET_VERSION variable to 171023 to match the modified client's version number.
 
+### Enabling testing
+While it is highly recommended to enable testing, if you would like to save compilation time, you'll want to comment out the enable_testing variable in CMakeVariables.txt.
+It is recommended that after building and if testing is enabled, to run `ctest` and make sure all the tests pass.
```
Comment (aronwk-aaron) [inline README.md:47]: aren't tests gtest, can you run gtest with ctest?

Comment (EmosewaMC) [reply README.md:47]: They are still run with ctest.   Gtest is just an infrastructure change.

File: dWorldServer/CMakeLists.txt [hunk 11]
```diff
@@ -1,6 +1,11 @@
...
-		"WorldServer.cpp")
+set(DWORLDSERVER_SOURCES
+	"ObjectIDManager.cpp"
+	"PerformanceManager.cpp"
+)
+
+add_library(dWorldServer ${DWORLDSERVER_SOURCES})
+add_executable(WorldServer "WorldServer.cpp")
+
+target_link_libraries(dWorldServer ${COMMON_LIBRARIES})
```
Comment (Jettford) [inline dWorldServer/CMakeLists.txt:9]: Really not certain if I am happy with this, if those things need to be used outside of WorldServer they should be taken out of the WorldServer directory.

File: tests/dGameTests/dComponentsTests/DestroyableComponentTests.cpp [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+		DestroyableComponent* destroyableComponent;
+		CBITSTREAM
+		uint32_t flags = 0;
+		void SetUp() override {
+			SetUpDependencies();
+			baseEntity = new Entity(15, GameDependenciesTest::info);
+			destroyableComponent = new DestroyableComponent(baseEntity);
+			baseEntity->AddComponent(COMPONENT_TYPE_DESTROYABLE, destroyableComponent);
+			// Initialize some values to be not default
+			destroyableComponent->SetMaxHealth(12345.0f);
```
Comment (Jettford) [inline tests/dGameTests/dComponentsTests/DestroyableComponentTests.cpp:20]: any specific reasons for these values? (this is just curiosity)

File: dWorldServer/CMakeLists.txt (hunk 11 above)
Comment (EmosewaMC) [reply dWorldServer/CMakeLists.txt:9]: They should be but that's outside the scope of this pr.  We needed to get things compiling and this is the simplest change to do so.  Moving these files would also help us resolve the duplicate objectidmanagers we have.  I see no reason we need to entirely separate managers.

File: tests/dGameTests/dComponentsTests/DestroyableComponentTests.cpp (hunk 12 above)
Comment (EmosewaMC) [reply tests/dGameTests/dComponentsTests/DestroyableComponentTests.cpp:20]: Just arbitrary values.  No meaning.

File: dWorldServer/CMakeLists.txt (hunk 11 above)
Comment (EmosewaMC) [reply dWorldServer/CMakeLists.txt:9]: We also need to do it like so so we can actually test these files in the future.  If left as is in main we could not test these files

**Comment by EmosewaMC** [comment]
I am closing this.  I don't believe enough people will care enough about this to use it and as such it serves no purpose to implement it as such.

**Comment by aronwk-aaron** [comment]
> I don't believe enough people will care enough about this to use it and as such it serves no purpose to implement it as such.

not true? The conversation that was had is not everything can, or should be tested via code, some things can only be verified via booting up the game and testing it. Even if a test cane be made to check everything, it's difficult, neigh impossible, to tell if everything works under the client via tests like these. 
As for the serializations and AMF tests, that's perfect se for tests since they work in an explicitly definable way.
For things like triggers (which sparked the conversation the other day) those are tricky since there's a lot of tomfoolery on the client side for them to work and we don't have an accurate physics engine server side to for sure know that we are doing things correctly from the client's perspective. 

TL;DR This should be re-opened and merged and used appropriately.
We can test programmatic things like serialization (especially since we are still working on that and it would help us prevent pushing completely incorrect serialization, which I am guilty of doing by accident), but automated testing is not the end all be all, and I speak from experience as automated testing is my literal job

**Review by Jettford — APPROVED** [review]

**Review by aronwk-aaron — APPROVED** [review]

**Review by Xiphoseer — APPROVED** [review]
LGTM

the redirecting all build types output into a single dir isn't ideal, but not nearly important enough to block this PR further.
